pallene/checker.lua

```diff
@@ -451,9 +452,42 @@ function Checker:check_stat(stat)
             self:add_local(stat.decl)
             self:check_stat(stat.block)
         end)
+    elseif tag == "ast.Stat.ForIn" then
+        local rhs = stat.exps
+        self:expand_function_returns(3, rhs)
+
+        local expected_ret_types = {}
+        for _ = 1, #stat.decls do
+            table.insert(expected_ret_types, types.T.Any())
+        end
+
+        local iteratorfn = self:check_exp_verify(rhs[1], types.T.Function({
+            types.T.Any(), types.T.Any()
+        }, expected_ret_types), "loop iterator")
+
+        if not rhs[2] then
+            type_error(stat.loc, "Missing state argument in for-in loop")
+        else
+            rhs[2] = self:check_exp_verify(rhs[2], types.T.Any(), "state variable")
+        end
+
+        rhs[3] = rhs[3] and self:check_exp_verify(rhs[3], types.T.Any(), "control variable")
+
+        if #stat.decls ~= #iteratorfn._type.ret_types then
+            type_error(stat.decls[1].loc, "expected %d values, but function returns %d",
+                       #stat.decls, #iteratorfn._type.ret_types)
+        end
+
+        self.symbol_table:with_block(function()
+            for i = 1, #stat.decls do
+                stat.decls[i]._type = iteratorfn._type.ret_types[i]
+                self:add_local(stat.decls[i])
```

Comment:
We should also handle the case where the programmer inserted a type annotation. That is, when `stat.decls[i].type` exists. (See check_initializer_exp). Perhaps use `types.consistent` to see if the type returned to the iterator can be assigned to the declared type?